### PR TITLE
Feature/utility code reorganize

### PR DESCRIPTION
**Pull request by xuzh2024**

### CODE REVIEW

File: indexer/utils/abi_setting.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+# -*- coding: utf-8 -*-
+"""
+Time    : 2024/10/12 下午2:10
+Author  : xuzh
+Project : hemera_indexer
+"""
+from indexer.utils.abi_code_utils import Event, Function
+
+# log event
+deposit_event = Event(
```
Comment (shanshuo0918) [inline indexer/utils/abi_setting.py:11]: WETH_DEPOSIT_EVENT

File: indexer/utils/abi_setting.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        "inputs": [
+            {"indexed": True, "name": "dst", "type": "address"},
+            {"indexed": False, "name": "wad", "type": "uint256"},
+        ],
+        "name": "Deposit",
+        "type": "event",
+    }
+)
+
+withdraw_event = Event(
```
Comment (shanshuo0918) [inline indexer/utils/abi_setting.py:23]: WETH_WITHDRAW_EVENT

File: indexer/utils/abi_setting.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        "inputs": [
+            {"indexed": True, "name": "src", "type": "address"},
+            {"indexed": False, "name": "wad", "type": "uint256"},
+        ],
+        "name": "Withdrawal",
+        "type": "event",
+    }
+)
+
+transfer_event = Event(
```
Comment (shanshuo0918) [inline indexer/utils/abi_setting.py:35]: ERC20_TRANSFER_EVENT

File: indexer/utils/abi_setting.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+            {"indexed": True, "name": "from", "type": "address"},
+            {"indexed": True, "name": "to", "type": "address"},
+            {"indexed": False, "name": "value", "type": "uint256"},
+        ],
+        "name": "Transfer",
+        "type": "event",
+    }
+)
+
+single_transfer_event = Event(
```
Comment (shanshuo0918) [inline indexer/utils/abi_setting.py:48]: ERC1155_SINGLE_TRANSFER_EVENT

File: indexer/utils/abi_setting.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+            {"indexed": True, "name": "to", "type": "address"},
+            {"indexed": False, "name": "id", "type": "uint256"},
+            {"indexed": False, "name": "value", "type": "uint256"},
+        ],
+        "name": "TransferSingle",
+        "type": "event",
+    }
+)
+
+batch_transfer_event = Event(
```
Comment (shanshuo0918) [inline indexer/utils/abi_setting.py:63]: ERC1155_BATCH_TRANSFER_EVENT

File: indexer/utils/abi_setting.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+            {"indexed": False, "name": "ids", "type": "uint256[]"},
+            {"indexed": False, "name": "values", "type": "uint256[]"},
+        ],
+        "name": "TransferBatch",
+        "type": "event",
+    }
+)
+
+# ABI function
+NAME_ABI_FUNCTION = Function(
```
Comment (shanshuo0918) [inline indexer/utils/abi_setting.py:79]: TOKEN_NAME_FUNCTION

File: indexer/utils/abi_setting.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+        "inputs": [],
+        "name": "name",
+        "outputs": [{"name": "", "type": "string"}],
+        "payable": False,
+        "stateMutability": "view",
+        "type": "function",
+    }
+)
+
+SYMBOL_ABI_FUNCTION = Function(
```
Comment (shanshuo0918) [inline indexer/utils/abi_setting.py:91]: TOKEN_SYMBOL_FUNCTION

File: indexer/utils/abi_setting.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+        "inputs": [],
+        "name": "symbol",
+        "outputs": [{"name": "", "type": "string"}],
+        "payable": False,
+        "stateMutability": "view",
+        "type": "function",
+    }
+)
+
+DECIMALS_ABI_FUNCTION = Function(
```
Comment (shanshuo0918) [inline indexer/utils/abi_setting.py:103]: TOKEN_DECIMALS_FUNCTION

File: indexer/utils/abi_setting.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+        "inputs": [],
+        "name": "decimals",
+        "outputs": [{"name": "", "type": "uint8"}],
+        "payable": False,
+        "stateMutability": "view",
+        "type": "function",
+    }
+)
+
+TOTAL_SUPPLY_ABI_FUNCTION = Function(
```
Comment (shanshuo0918) [inline indexer/utils/abi_setting.py:115]: TOKEN_TOTAL_SUPPLY_FUNCTION

File: indexer/utils/abi_setting.py [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+        "inputs": [],
+        "name": "totalSupply",
+        "outputs": [{"name": "", "type": "uint256"}],
+        "payable": False,
+        "stateMutability": "view",
+        "type": "function",
+    }
+)
+
+TOTAL_SUPPLY_WITH_ID_ABI_FUNCTION = Function(
```
Comment (shanshuo0918) [inline indexer/utils/abi_setting.py:127]: TOKEN_TOTAL_SUPPLY_WITH_ID_FUNCTION

File: indexer/utils/abi_setting.py [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+        "inputs": [{"name": "id", "type": "uint256"}],
+        "name": "totalSupply",
+        "outputs": [{"name": "", "type": "uint256"}],
+        "payable": False,
+        "stateMutability": "view",
+        "type": "function",
+    }
+)
+
+TOKEN_SUPPLY_ABI_FUNCTION = Function(
```
Comment (shanshuo0918) [inline indexer/utils/abi_setting.py:139]: Where does this come from?

File: indexer/utils/abi_setting.py [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+        "inputs": [],
+        "name": "tokenSupply",
+        "outputs": [{"name": "", "type": "uint256"}],
+        "payable": False,
+        "stateMutability": "view",
+        "type": "function",
+    }
+)
+
+OWNER_OF_ABI_FUNCTION = Function(
```
Comment (shanshuo0918) [inline indexer/utils/abi_setting.py:151]: ERC721_OWNER_OF_FUNCTION

File: indexer/utils/abi_setting.py [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+        "inputs": [{"name": "tokenId", "type": "uint256"}],
+        "name": "ownerOf",
+        "outputs": [{"name": "owner", "type": "address"}],
+        "payable": False,
+        "stateMutability": "view",
+        "type": "function",
+    }
+)
+
+TOKEN_URI_ABI_FUNCTION = Function(
```
Comment (shanshuo0918) [inline indexer/utils/abi_setting.py:163]: ERC721_TOKEN_URI_FUNCTION

File: indexer/utils/abi_setting.py [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+        "inputs": [{"name": "tokenId", "type": "uint256"}],
+        "name": "tokenURI",
+        "outputs": [{"name": "uri", "type": "address"}],
+        "payable": False,
+        "stateMutability": "view",
+        "type": "function",
+    }
+)
+
+URI_ABI_FUNCTION = Function(
```
Comment (shanshuo0918) [inline indexer/utils/abi_setting.py:175]: ERC1155_MULTIPLE_TOKEN_URI_FUNCTION

File: indexer/utils/abi_setting.py [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+        "inputs": [{"name": "_owner", "type": "address"}],
+        "name": "balanceOf",
+        "outputs": [{"name": "balance", "type": "uint256"}],
+        "payable": False,
+        "stateMutability": "view",
+        "type": "function",
+    }
+)
+
+BALANCE_OF_WITH_TOKEN_ID_ABI_FUNCTION = Function(
```
Comment (shanshuo0918) [inline indexer/utils/abi_setting.py:199]: ERC1155_TOKEN_ID_BALANCE_OF_FUNCTION

File: indexer/utils/abi_setting.py (hunk 3 above)
Comment (shanshuo0918) [reply indexer/utils/abi_setting.py:35]: ERC721_TRANSFER_EVENT

Comment (shanshuo0918) [reply indexer/utils/abi_setting.py:35]: same value

File: indexer/utils/abi_setting.py [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+        "inputs": [{"name": "tokenId", "type": "uint256"}],
+        "name": "uri",
+        "outputs": [{"name": "uri", "type": "address"}],
+        "payable": False,
+        "stateMutability": "view",
+        "type": "function",
+    }
+)
+
+BALANCE_OF_ABI_FUNCTION = Function(
```
Comment (shanshuo0918) [inline indexer/utils/abi_setting.py:187]: ERC20_BALANCE_OF_FUNCTION

File: indexer/utils/abi_code_utils.py [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+    codec,
+    event_log_abi_to_topic,
+    function_abi_to_4byte_selector_str,
+    get_types_from_abi_type_list,
+)
+
+abi_codec = ABICodec(eth_abi.registry.registry)
+
+
+class Event:
```
Comment (shanshuo0918) [inline indexer/utils/abi_code_utils.py:41]: Add some unit test

File: indexer/utils/enrich.py [hunk 18]
```diff
@@ -1,269 +0,0 @@
-import itertools
```
Comment (shanshuo0918) [inline indexer/utils/enrich.py:1]: We don't use this functions anymore?

File: indexer/domain/log.py [hunk 19]
```diff
@@ -40,21 +40,21 @@ def from_rpc(log_dict: dict, block_timestamp=None, block_hash=None, block_number
             topic3=topics[3] if len(topics) > 3 else None,
         )
 
-    def get_bytes_topics(self) -> HexBytes:
+    def get_bytes_topics(self) -> bytes:
```
[the remark this answers is not in the document]
Comment (xuzh2024) [reply indexer/domain/log.py:43]: HexBytes inherits from bytes and is primarily used in the Web3.py library. HexBytes and bytes are compatible with each other when representing Ethereum-related binary data. 
HexBytes provides a more user-friendly way to print and represent hexadecimal data. For example, when you print a HexBytes object, it is displayed as a hexadecimal string, whereas a bytes object is displayed as a sequence of bytes.
In addition to this, HexBytes may have some additional methods or properties that are not available on normal bytes objects. Things like value checking during initialization, and so on.

File: indexer/modules/bridge/bedrock/parser/function_parser/finalize_bridge_erc20.py [hunk 20]
```diff
@@ -61,7 +62,7 @@
 def decode_function(input_data: bytes) -> BridgeRemoteFunctionCallInfo:
     assert len(input_data) >= 4
     print(function_abi_to_4byte_selector_str(FINALIZE_BRIDGE_ERC20))
-    assert function_abi_to_4byte_selector_str(FINALIZE_BRIDGE_ERC20) == "0x" + input_data[:4].hex().lower()
+    assert function_abi_to_4byte_selector_str(FINALIZE_BRIDGE_ERC20) == bytes_to_hex_str(input_data[:4]).lower()
```
[the remark this answers is not in the document]
Comment (xuzh2024) [reply indexer/modules/bridge/bedrock/parser/function_parser/finalize_bridge_erc20.py:65]: function_abi_to_4byte_selector_str is encapsulated by function_abi_to_function.get_signature.
This code is in modules, so its abi calls aren't unified yet.

File: indexer/modules/bridge/bedrock/parser/function_parser/finalize_bridge_erc721.py [hunk 21]
```diff
@@ -61,7 +62,7 @@
 def decode_function(input_data: bytes) -> BridgeRemoteFunctionCallInfo:
     assert len(input_data) >= 4
     print(function_abi_to_4byte_selector_str(FINALIZE_BRIDGE_ERC721))
-    assert function_abi_to_4byte_selector_str(FINALIZE_BRIDGE_ERC721) == "0x" + input_data[:4].hex().lower()
+    assert function_abi_to_4byte_selector_str(FINALIZE_BRIDGE_ERC721) == bytes_to_hex_str(input_data[:4])
```
[the remark this answers is not in the document]
Comment (xuzh2024) [reply indexer/modules/bridge/bedrock/parser/function_parser/finalize_bridge_erc721.py:65]: Probably for exception handling, I'm going to replace assert with conditional + throw exception

File: indexer/utils/enrich.py (hunk 18 above)
Comment (xuzh2024) [reply indexer/utils/enrich.py:1]: Yes.
The enrich info code from earlier is incorporated into the dataclass initialization.

File: indexer/utils/abi_setting.py (hunk 11 above)
Comment (xuzh2024) [reply indexer/utils/abi_setting.py:139]: I can't remember where I saw this abi function, it doesn't seem to be in use right now, I'll remove this code in a subsequent commit

**Review by shanshuo0918 — APPROVED** [review]

**Pull request merged**